File: experiments/adrianco/experiment-5/runs/language=python_model=claude-opus-4-7_tooling=beads/rep2/src/soccer_mcp/players.py
"""FIFA player queries.

Search the FIFA player dump by name, nationality, club, and position; return a
short dict per player rather than a pandas row so consumers don't need pandas
in their dependency tree. All comparisons go through the normalised
``name_norm`` / ``club_norm`` / ``nationality_norm`` columns added by
``data._load_fifa`` so a query for "São Paulo" matches a club tagged
"Sao Paulo FC", and a name search for "neymar" finds "Neymar Jr".
"""

from __future__ import annotations

import unicodedata
from typing import Any

import pandas as pd

from soccer_mcp.data import SoccerData, normalize_team_name


def _strip_accents(text: str) -> str:
    return "".join(c for c in unicodedata.normalize("NFD", text) if unicodedata.category(c) != "Mn")


def _player_dict(row: pd.Series) -> dict[str, Any]:
    def get(col: str, default=None):
        if col not in row.index:
            return default
        val = row[col]
        if pd.isna(val):
            return default
        return val

    return {
        "id": int(get("ID")) if get("ID") is not None else None,
        "name": get("Name"),
        "age": int(get("Age")) if get("Age") is not None else None,
        "nationality": get("Nationality"),
        "overall": int(get("Overall")) if get("Overall") is not None else None,
        "potential": int(get("Potential")) if get("Potential") is not None else None,
        "club": get("Club"),
        "position": get("Position"),
        "jersey_number": int(get("Jersey Number")) if get("Jersey Number") is not None else None,
        "height": get("Height"),
        "weight": get("Weight"),
        "preferred_foot": get("Preferred Foot"),
    }


def search_players_by_name(data: SoccerData, name: str, limit: int = 25) -> list[dict]:
    """Substring match on player name (accent-insensitive)."""
    needle = _strip_accents(name).lower().strip()
    if not needle:
        return []
    df = data.fifa
    matches = df[df["name_norm"].str.contains(needle, na=False, regex=False)]
    matches = matches.sort_values("Overall", ascending=False).head(limit)
    return [_player_dict(row) for _, row in matches.iterrows()]


def players_by_nationality(data: SoccerData, nationality: str, limit: int = 25) -> list[dict]:
    """All players from a given country, sorted by overall rating."""
    target = nationality.strip().lower()
    df = data.fifa[data.fifa["nationality_norm"] == target]
    df = df.sort_values("Overall", ascending=False).head(limit)
    return [_player_dict(row) for _, row in df.iterrows()]


def players_by_club(
    data: SoccerData,
    club: str,
    position: str | None = None,
    limit: int = 50,
) -> list[dict]:
    """Players at a club, optionally filtered to a single position."""
    club_norm = normalize_team_name(club)
    df = data.fifa[data.fifa["club_norm"].str.contains(club_norm, na=False, regex=False)]
    if position:
        df = df[df["Position"].fillna("").str.upper() == position.upper()]
    df = df.sort_values("Overall", ascending=False).head(limit)
    return [_player_dict(row) for _, row in df.iterrows()]


def top_players(
    data: SoccerData,
    nationality: str | None = None,
    position: str | None = None,
    limit: int = 10,
) -> list[dict]:
    """Highest-rated players in scope; defaults to global top-N."""
    df = data.fifa
    if nationality:
        df = df[df["nationality_norm"] == nationality.strip().lower()]
    if position:
        df = df[df["Position"].fillna("").str.upper() == position.upper()]
    df = df.sort_values("Overall", ascending=False).head(limit)
    return [_player_dict(row) for _, row in df.iterrows()]


def brazilian_players_by_club_summary(data: SoccerData, limit: int = 20) -> list[dict]:
    """Per-club counts and avg rating among Brazilian players."""
    df = data.fifa[data.fifa["nationality_norm"] == "brazil"]
    grouped = df.groupby("Club").agg(
        players=("Name", "count"),
        avg_rating=("Overall", "mean"),
    )
    grouped = grouped.sort_values("players", ascending=False).head(limit)
    return [
        {"club": club, "players": int(row["players"]), "avg_rating": round(float(row["avg_rating"]), 1)}
        for club, row in grouped.iterrows()
    ]

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=none/rep3/src/queries/players.ts
import { Player } from '../types.js';
import { stripAccents, normalizeTeam, keyMatches } from '../normalize.js';

export interface PlayerFilter {
  /** Substring of player name (case/diacritic-insensitive). */
  name?: string;
  /** Nationality, exact (case-insensitive, accent-insensitive). */
  nationality?: string;
  /** Club name; uses team normalization so "Flamengo" matches "CR Flamengo". */
  club?: string;
  /** Position code (e.g. "ST", "LW"). Case-insensitive substring. */
  position?: string;
  /** Minimum FIFA Overall rating. */
  minOverall?: number;
  /** Minimum age (inclusive). */
  minAge?: number;
  /** Maximum age (inclusive). */
  maxAge?: number;
  limit?: number;
}

function cmpInsensitive(a: string, b: string): boolean {
  return stripAccents(a.toLowerCase()) === stripAccents(b.toLowerCase());
}

function containsInsensitive(haystack: string, needle: string): boolean {
  return stripAccents(haystack.toLowerCase()).includes(
    stripAccents(needle.toLowerCase()),
  );
}

export function findPlayers(players: Player[], filter: PlayerFilter): Player[] {
  const clubKey = filter.club ? normalizeTeam(filter.club) : '';
  let out = players.filter((p) => {
    if (filter.name && !containsInsensitive(p.name, filter.name)) return false;
    if (
      filter.nationality &&
      !(p.nationality && cmpInsensitive(p.nationality, filter.nationality))
    )
      return false;
    if (clubKey) {
      if (!p.clubKey) return false;
      if (!keyMatches(p.clubKey, clubKey)) return false;
    }
    if (filter.position) {
      if (!p.position) return false;
      if (!containsInsensitive(p.position, filter.position)) return false;
    }
    if (filter.minOverall != null && (p.overall ?? 0) < filter.minOverall)
      return false;
    if (filter.minAge != null && (p.age ?? 0) < filter.minAge) return false;
    if (filter.maxAge != null && (p.age ?? 999) > filter.maxAge) return false;
    return true;
  });

  // Highest-rated first by default.
  out.sort((a, b) => (b.overall ?? 0) - (a.overall ?? 0));
  if (filter.limit != null && filter.limit > 0) out = out.slice(0, filter.limit);
  return out;
}

export interface ClubSummary {
  club: string;
  count: number;
  averageOverall: number;
  topPlayer?: Player;
}

/**
 * Aggregate players by their club. Useful for "Brazilian players at Brazilian
 * clubs" style summaries.
 */
export function playersByClub(players: Player[]): ClubSummary[] {
  const byKey = new Map<string, { club: string; players: Player[] }>();
  for (const p of players) {
    if (!p.clubKey || !p.club) continue;
    const entry = byKey.get(p.clubKey) ?? { club: p.club, players: [] };
    entry.players.push(p);
    byKey.set(p.clubKey, entry);
  }
  const out: ClubSummary[] = [];
  for (const { club, players: ps } of byKey.values()) {
    const overall = ps
      .map((p) => p.overall ?? 0)
      .filter((n) => n > 0);
    const avg = overall.length
      ? overall.reduce((a, b) => a + b, 0) / overall.length
      : 0;
    const top = ps
      .slice()
      .sort((a, b) => (b.overall ?? 0) - (a.overall ?? 0))[0];
    out.push({
      club,
      count: ps.length,
      averageOverall: avg,
      topPlayer: top,
    });
  }
  return out.sort((a, b) => b.averageOverall - a.averageOverall);
}
